### PR TITLE
Case 20613: Avatar Exporter v0.2 - copy external textures

### PR DESCRIPTION
**Pull request by dback2**
https://highfidelity.manuscript.com/f/cases/20613/

New Avatar Exporter package:
https://github.com/highfidelity/hifi/raw/348ac3167ffb8ba20a25e86b5a6a69407598d36d/tools/unity-avatar-exporter/avatarExporter.unitypackage

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17151/HighFidelity-Beta-PR14881-130df1b.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17151/HighFidelity-Beta-PR14881-130df1b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17151/HighFidelity-Beta-Interface-PR14881-130df1b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17151/HighFidelity-Beta-PR14881-130df1b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17151/HighFidelity-Beta-Interface-PR14881-130df1b.exe

**Comment by jakram** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17220/HighFidelity-Beta-PR14881-130df1b.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17220/HighFidelity-Beta-PR14881-130df1b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17220/HighFidelity-Beta-Interface-PR14881-130df1b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17220/HighFidelity-Beta-PR14881-130df1b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17220/HighFidelity-Beta-Interface-PR14881-130df1b.exe

**Comment by jakram** [comment]
Currently testing

**Comment by jakram** [comment]
Testing paused

**Comment by jakram** [comment]
Passes QA testing

File: tools/unity-avatar-exporter/Assets/Editor/AvatarExporter.cs [hunk 1]
```diff
@@ -479,6 +479,26 @@ class BoneTreeNode {
...
+            if (textureObject != null) {
+                texturePaths.Add(dependencyPath);
+            }
+        }
+
+        // copy the found dependency textures from the local asset folder to the textures folder in the target export project
+        foreach (string texturePath in texturePaths) {
+            string textureName = Path.GetFileName(texturePath);
+            string targetPath = texturesDirectory + "\\" + textureName;
+            File.Copy(texturePath, targetPath);
```
Comment (huffman) [inline tools/unity-avatar-exporter/Assets/Editor/AvatarExporter.cs]: Should we add error handling to this? I can imagine copying failing in certain cases.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17307/HighFidelity-Beta-PR14881-41abacb.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17307/HighFidelity-Beta-PR14881-41abacb.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17307/HighFidelity-Beta-Interface-PR14881-41abacb.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17307/HighFidelity-Beta-PR14881-41abacb.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17307/HighFidelity-Beta-Interface-PR14881-41abacb.exe

File: tools/unity-avatar-exporter/Assets/Editor/AvatarExporter.cs [hunk 2]
```diff
@@ -455,12 +449,20 @@ class BoneTreeNode {
             }
 
             // write out a new fst file in place of the old file
-            WriteFST(exportFstPath, projectName);
+            if (!WriteFST(exportFstPath, projectName)) {
+                return;
+            }
+            
+            // copy any external texture files to the project's texture directory that are considered dependencies of the model
+            string texturesDirectory = Path.GetDirectoryName(exportFstPath) + "\\textures";
```
Comment (huffman) [inline tools/unity-avatar-exporter/Assets/Editor/AvatarExporter.cs]: Can we put the textures path in a string and reuse it in the couple places where it is used?

Comment (dback2) [reply tools/unity-avatar-exporter/Assets/Editor/AvatarExporter.cs]: Added a GetTextureDirectory from base path

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17336/HighFidelity-Beta-PR14881-348ac31.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17336/HighFidelity-Beta-PR14881-348ac31.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17336/HighFidelity-Beta-Interface-PR14881-348ac31.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17336/HighFidelity-Beta-PR14881-348ac31.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17336/HighFidelity-Beta-Interface-PR14881-348ac31.exe

**Pull request merged**